File: 1497-check-if-array-pairs-are-divisible-by-k/1497-check-if-array-pairs-are-divisible-by-k.py
class Solution:
    def canArrange(self, arr: List[int], k: int) -> bool:
        
        if len(arr)%2 == 1:
            return False
        
        fmap = {}
        for ele in arr:
            rem = ((ele%k) + k) % k;
            fmap[rem] = fmap.get(rem, 0) + 1
        
        
        for ele in arr:
            rem = ((ele%k) + k) % k;
            
            if rem == 0:
                if fmap[rem] % 2 == 1:
                    return False
            else:
                if fmap[rem] != fmap.get(k-rem, 0):
                    return False
        
        return True
    

File: 1497-check-if-array-pairs-are-divisible-by-k/1497-check-if-array-pairs-are-divisible-by-k.java
class Solution {
    public boolean canArrange(int[] arr, int k) {
        
        if(arr.length % 2 == 1) return false;
        
        Map<Integer, Integer> fmap = new HashMap<>();
        for(int a : arr){
            int rem = ((a%k)+k)%k;
            fmap.put(rem, fmap.getOrDefault(rem, 0) + 1);
        }
        
        for(int a : arr){
            
            int rem = ((a%k)+k)%k;
            
            if(rem == 0){
                if(fmap.get(rem) % 2 == 1) return false;
            }
            else{
                int xFreq = fmap.get(rem);
                int k_xFreq = fmap.getOrDefault(k-rem, 0);
                
                if(xFreq != k_xFreq) return false;
            }
        }
        
        return true;
    }
}